pallets/rewards/src/lib.rs

```diff
@@ -0,0 +1,291 @@
+            signature.verify(msg.as_slice(), itn_address),
+            Error::<T>::InvalidSignature
+        );
+        Ok(())
+    }
+
+    pub fn base_claim_itn_reward(
+        reward_address: T::AccountId,
+        itn_address: T::AccountId,
+        signature: T::OffChainSignature,
+    ) -> DispatchResult {
+        <ItnRewards<T>>::try_mutate(&itn_address, |reward| {
+            match reward {
+                // Unclaimed. Attempt to claim.
+                Some(ItnRewardStatus::Unclaimed(amount)) => {
+                    // `amount` and `bonded_amount` are equal in value but different types.
+                    // `deposit_amount` is 1 POLY more because we bond `bonded_amount`
+                    // and we don't want all of the user's poly bonded.
+                    let amount = *amount;
+                    let (bonded_amount, deposit_amount) = Self::convert_balance(amount)?;
+
+                    // Verify the signature.
+                    Self::verify_itn_sig(&reward_address, &itn_address, &signature)?;
+
+                    // `with_transaction` is required because we must undo the transfer if bonding fails.
+                    with_transaction(|| {
+                        // Transfer `deposit_amount` from the rewards treasury to the `reward_address`.
+                        T::Currency::transfer(
+                            &Self::account_id(),
+                            &reward_address,
+                            deposit_amount,
+                            ExistenceRequirement::AllowDeath,
+                        )?;
+
+                        // Bond additional `bonded_amount` for `reward_address`.
+                        let origin = RawOrigin::Signed(reward_address.clone()).into();
+                        if <Staking<T>>::bonded(&reward_address).is_some() {
+                            <Staking<T>>::bond_extra(origin, bonded_amount)
+                        } else {
+                            <Staking<T>>::bond(
+                                origin,
+                                T::Lookup::unlookup(reward_address.clone()),
+                                bonded_amount,
+                                RewardDestination::Staked,
+                            )
+                        }
+                    })?;
+
+                    // Set the reward to claimed.
+                    *reward = Some(ItnRewardStatus::Claimed);
+                    Self::deposit_event(Event::<T>::ItnRewardClaimed(reward_address, amount));
+                    Ok(())
+                }
+                // Already Claimed.
+                Some(ItnRewardStatus::Claimed) => Err(Error::<T>::ItnRewardAlreadyClaimed.into()),
+                // Unknown Address.
+                None => Err(Error::<T>::UnknownItnAddress.into()),
+            }
+        })
+    }
+
+    pub fn base_set_itn_reward_status(itn_address: &T::AccountId, status: ItnRewardStatus) {
+        <ItnRewards<T>>::insert(itn_address, status);
```

Comment:
Let's emit an event?

Reply:
Also move the `ensure_root` check into this function.  Functions with `base_` prefix should do checks.  Functions without ensure checks should be prefixed `unverified_`.

Reply:
`ensure_root` doesn't seem like a check that should be required in the base function. It's not a requirement to be root origin to call the base function and so why enforce the caller pass a root origin?

Reply:
The idea with `base_` was to separate more logic out of the macro to a) apply rustfmt more, b) improve coverage, but contain all of the logic an extrinsic needed. The idea with `unverified_` functions was for the cases where the logic wasn't self contained and some verification steps were omitted. So if you want to keep `ensure_root` outside then let's rename the function.